src/dwt.rs

```diff
@@ -0,0 +1,219 @@
+    pub fn stopwatch<'i>(&self, times: &'i mut [u32]) -> StopWatch<'i> {
+        StopWatch::new(times, self.clocks.hclk())
+    }
+    /// Measure cycles it takes to execute f
+    pub fn measure<F: FnOnce()>(&self, f: F) -> ClockDuration {
+        let mut times: [u32; 2] = [0; 2];
+        let mut sw = self.stopwatch(&mut times);
+        f();
+        sw.lap().lap_time(1).unwrap()
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Delay {
+    clock: Hertz,
+}
+impl Delay {
+    /// Delay for ClockDuration::ticks
+    pub fn delay(duration: ClockDuration) {
+        let ticks = duration.ticks as u64;
+        Delay::delay_ticks(DWT::get_cycle_count(), ticks);
+    }
+    /// Delay ticks
+    /// NOTE DCB and DWT need to be set up for this to work, so it is private
+    fn delay_ticks(mut start: u32, ticks: u64) {
+        if ticks < (u32::MAX / 2) as u64 {
+            // Simple delay
+            let ticks = ticks as u32;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < ticks {}
+        } else if ticks <= u32::MAX as u64 {
+            // Try to avoid race conditions by limiting delay to u32::MAX / 2
+            let mut ticks = ticks as u32;
+            ticks -= u32::MAX / 2;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < u32::MAX / 2 {}
+            start -= u32::MAX / 2;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < ticks {}
+        } else {
+            // Delay for ticks, then delay for rest * u32::MAX
+            let mut rest = (ticks >> 32) as u32;
+            let ticks = (ticks & u32::MAX as u64) as u32;
+            loop {
+                while (DWT::get_cycle_count().wrapping_sub(start)) < ticks {}
+                if rest == 0 {
+                    break;
+                }
+                rest -= 1;
+                while (DWT::get_cycle_count().wrapping_sub(start)) > ticks {}
+            }
+        }
+    }
+}
+
+// Implement DelayUs/DelayMs for various integer types
+impl DelayUs<u64> for Delay {
+    fn delay_us(&mut self, us: u64) {
+        // Convert us to ticks
+        let start = DWT::get_cycle_count();
+        let ticks = (us * self.clock.0 as u64) / 1_000_000;
+        Delay::delay_ticks(start, ticks);
+    }
+}
+impl DelayMs<u64> for Delay {
+    fn delay_ms(&mut self, ms: u64) {
+        // Convert ms to ticks
+        let start = DWT::get_cycle_count();
+        let ticks = (ms * self.clock.0 as u64) / 1_000;
+        Delay::delay_ticks(start, ticks);
+    }
+}
+macro_rules! impl_DelayIntT {
+    (for $($t:ty),+) => {$(
+        impl DelayUs<$t> for Delay {
+            fn delay_us(&mut self, us: $t) {
+                self.delay_us(us as u64);
+            }
+        }
+        impl DelayMs<$t> for Delay {
+            fn delay_ms(&mut self, ms: $t) {
+                self.delay_ms(ms as u64);
+            }
+        }
+    )*}
+}
+impl_DelayIntT!(for usize,  u32, u16, u8);
```

Comment:
You can avoid this macro by impl<T: Into<u64>> (maybe not for usize)

Reply:
changed to impl into